### PR TITLE
Template and script for W->Z'->3mu+nu signal generation

### PR DESCRIPTION
**Pull request by Nik-Menendez**
This pull request is for the official generation of Z' production in W->3mu+nu final states. This contains the template cards in bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/ and a script to generate all the mass points for each W charge. I'm aware the run card is outdated, let me know what to change so I can get it up to date.

### CODE REVIEW

**Comment by Nik-Menendez** [comment]
Please review this @sihyunjeon

**Comment by Saptaparna** [comment]
Why have you set the BW-cutoff to 15000? The simulation is not really reliable with such a large value of the BWCutOff.

**Comment by Nik-Menendez** [comment]
I just used all the same settings located in this run card which was already available in genproductions: https://github.com/cms-sw/genproductions/blob/master/bin/MadGraph5_aMCatNLO/cards/production/2017/WplusZp_to_3mu/Zp_to_3mu_M15/Zp_to_3mu_M15_run_card.dat

If there are better settings or other things I should change let me know and I can change them to whatever is best.

**Comment by Nik-Menendez** [comment]
Is there any update on this? Is there anyone I can message to get more assistance with generating this MC?

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_customizecards.dat [hunk 1]
```diff
@@ -0,0 +1,9 @@
+set param_card mass 999888 ZPMASS
+set param_card ZPINPUTS 3 1.000000e-01
+set param_card ZPINPUTS 4 1.000000e-01
+set param_card ZPINPUTS 5 -1.000000e-01
+set param_card ZPINPUTS 6 -1.000000e-01
+set param_card DECAY 999888 Auto
+set param_card DECAY 23 Auto
+set param_card DECAY 24 Auto
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_customizecards.dat:8]: lines 7 and 8 may not needed as they are sm particles

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#*    The MadGraph5_aMC@NLO Development Team - Find us at   *
+#*    https://server06.fynu.ucl.ac.be/projects/madgraph     *
+#*                                                          *
+#************************************************************
+#*                                                          *
+#*               Command File for MadGraph5_aMC@NLO         *
+#*                                                          *
+#*     run as ./bin/mg5_aMC  filename                       *
+#*                                                          *
+#************************************************************
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat:22]: could you please remove the banner

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+set group_subprocesses Auto
+set ignore_six_quark_processes False
+set loop_optimized_output True
+set loop_color_flows False
+set gauge unitary
+set complex_mass_scheme False
+set max_npoint_for_channel 0
+import model sm
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat:30]: line 30 is not needed

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import model Leptophilic_UFO
+define p = g u c d s b u~ c~ d~ s~ b~
+define j = g u c d s b u~ c~ d~ s~ b~
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_proc_card.dat:33]: i noticed that all the quarks are massive in the Leptophilic_UFO model, we may need some discussion on it @Saptaparna @GurpreetSinghChahal @sihyunjeon @agrohsje

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+#*********************************************************************
+# Beam polarization from -100 (left-handed) to 100 (right-handed)    *
+#*********************************************************************
+        0     = polbeam1 ! beam polarization for beam 1
+        0     = polbeam2 ! beam polarization for beam 2
+#*********************************************************************
+# PDF CHOICE: this automatically fixes also alpha_s and its evol.    *
+#*********************************************************************
+ 'lhapdf'    = pdlabel     ! PDF set                                  
+ 263400    = lhaid     ! if pdlabel=lhapdf, this is the lhapdf number
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:52]: replace line 52 with "$DEFAULT_PDF_SETS = lhaid", and add "$DEFAULT_PDF_MEMBERS = reweight_PDF" right after line 52

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+ 1        = scalefact        ! scale factor for event-by-event scales
+#*********************************************************************
+# Matching - Warning! ickkw > 1 is still beta
+#*********************************************************************
+  0	= ickkw ! 0 no matching, 1 MLM, 2 CKKW matching
+  1	= highestmult ! for ickkw=2, highest mult group
+  1	= ktscheme ! for ickkw=1, 1 Durham kT, 2 Pythia pTE
+  1.0	= alpsfact ! scale factor for QCD emission vx
+  F	= chcluster ! cluster only according to channel diag
+  T	= pdfwgt ! for ickkw=1, perform pdf reweighting
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:71]: change line 71 to "F = pdfwgt"

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+ 263401   = PDF_set_min      ! First of the error PDF sets 
+ 263500   = PDF_set_max      ! Last of the error PDF sets
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:73]: remove these two lines

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+  4	= asrwgtflavor ! highest quark flavor for a_s reweight
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:74]: you use 5 flavor in the proc_card, so change here to 5

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+#*********************************************************************
+# KT DURHAM CUT                                                      *
+#*********************************************************************
+ -1    =  ktdurham        
+ 0.4  =  dparameter 
+#*********************************************************************
+# maximal pdg code for quark to be considered as a light jet         *
+# (otherwise b cuts are applied)                                     *
+#*********************************************************************
+ 4 = maxjetflavor    ! Maximum jet pdg code
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:259]: line 259 also keep it consistent with flavor scheme

File: bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+# Automatic ptj and mjj cuts if xqcut > 0
+# (turn off for VBF and single top processes)
+#**********************************************************
+   T  = auto_ptj_mjj  ! Automatic setting of ptj and mjj
+#**********************************************************
+#                                                                    
+#**********************************
+# BW cutoff (M+/-bwcutoff*Gamma)
+#**********************************
+  15000  = bwcutoff      ! (M+/-bwcutoff*Gamma)
```
Comment (menglu21) [inline bin/MadGraph5_aMCatNLO/cards/production/13TeV/ZpTo3mu/PROCESSNAME_run_card.dat:91]: there is always discussion on the bwcutoff, from the MG convention, when the decay chain is used (just as your process), the bwcutoff will be used to define the on-shell resonance, in which a narrow width limit is required. Very large bwcutoff will spoil the narrow width limit and should be avoided.

**Comment by Nik-Menendez** [comment]
The requested changes were made including reducing the bwcutoff to 15 instead. The only change not addressed is the possible massive quarks in the Leptophilic_UFO model that may require discusison.

**Comment by menglu21** [comment]
> The requested changes were made including reducing the bwcutoff to 15 instead. The only change not addressed is the possible massive quarks in the Leptophilic_UFO model that may require discusison.

Hi, i added two restriction cards for the model, one is to set udsc and ele/muon to be massless (the default), and the other one with addition massless b quark. Therefore there will be three options in this model,  "Leptophilic_UFO-full, Leptophilic_UFO-no_b_mass,  Leptophilic_UFO", the motivation of massless quark is to set the MC generation consistent with the PDF, the final difference on the xs is small (i checked this and you could confirm). 

N.B:
1. in both restriction cards, all of the ZP couplings are set to nonzero (in order to make you could set it on-the-fly), so please add 
"set param_card ZPINPUTS 1 0.000000e+00"
"set param_card ZPINPUTS 2 0.000000e+00"
in your template customized card if you want to turn off the Zp to ee coupling.

2. change "import model Leptophilic_UFO" to "import model Leptophilic_UFO-no_b_mass" if you want to use 5f scheme

please let me know if there is any issue when using the updated model.

**Comment by Nik-Menendez** [comment]
Hello, I changed the model to Leptophilic_UFO-no_b_mass and set the electron couplings to 0. I tested it and and it works. Is there anything else I need to do?

**Comment by menglu21** [comment]
Hi, i don't have further comment then, let me ping other and see if they have any. @Saptaparna @GurpreetSinghChahal @mlizzo please have a check, and merge if you are fine with them.

**Comment by mlizzo** [comment]
Hi @menglu21 , it seems to me that everything has been addressed, I'm fine with merging this PR

**Comment by mlizzo** [comment]
Done (see commit https://github.com/cms-sw/genproductions/commit/d985e0af17a76e067f067222cb5f7b74ac8a286e)

**Pull request merged**